File: 백준/Gold/4179. 불！/불！.py
import sys
import math
from collections import deque

input = sys.stdin.readline

R, C = map(int, input().split())
board = [[-1 for _ in range(C + 2)] for _ in range(R + 2)]
visited = [[False for _ in range(C + 2)] for _ in range(R + 2)]

MAX_INT = math.inf

dx = [-1, 0, 1, 0]
dy = [0, 1, 0, -1]

que = deque([])
x, y = 0, 9

for row in range(1, R + 1):
    line = input().strip()
    for col in range(1, len(line) + 1):
        s = line[col - 1]

        if s == "J":
            x, y = row, col
            board[row][col] = MAX_INT
        elif s == ".":
            board[row][col] = MAX_INT
        elif s == "F":
            que.append([row, col])
            board[row][col] = 0
        else:
            board[row][col] = -1

while que:
    cur_x, cur_y = que.popleft()

    for i in range(4):
        next_x, next_y = cur_x + dx[i], cur_y + dy[i]

        if board[next_x][next_y] == MAX_INT:
            board[next_x][next_y] = board[cur_x][cur_y] + 1
            que.append([next_x, next_y])

que.append([x, y, 0])
visited[x][y] = True

while que:
    cur_x, cur_y, time = que.popleft()

    if cur_x == 1 or cur_y == 1 or cur_x == R or cur_y == C:
        print(time + 1)
        exit()

    for i in range(4):
        next_x, next_y = cur_x + dx[i], cur_y + dy[i]

        if not visited[next_x][next_y] and board[next_x][next_y] > time + 1:
            visited[next_x][next_y] = True
            que.append([next_x, next_y, time + 1])

print("IMPOSSIBLE")

File: 백준/Gold/4179. 불！/불！.js
let input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n");

const [R, C] = input[0].split(" ").map((e) => +e);
board = Array.from({ length: R + 2 }, () => Array(C + 2).fill(0));
visited = Array.from({ length: R + 2 }, () => Array(C + 2).fill(false));

const MAX_INT = Infinity;

dx = [-1, 0, 1, 0];
dy = [0, 1, 0, -1];

const que = [];
let [x, y] = [0, 9];

for (let row = 1; row < R + 1; row++) {
  line = input[row];
  for (let col = 1; col < line.length + 1; col++) {
    let s = line[col - 1];

    if (s === "J") {
      x = row;
      y = col;
      board[row][col] = MAX_INT;
    } else if (s === ".") {
      board[row][col] = MAX_INT;
    } else if (s === "F") {
      que.push([row, col]);
      board[row][col] = 0;
    } else {
      board[row][col] = -1;
    }
  }
}

while (que.length > 0) {
  const [cur_x, cur_y] = que.shift();

  for (let i = 0; i < 4; i++) {
    const next_x = cur_x + dx[i];
    const next_y = cur_y + dy[i];

    if (board[next_x][next_y] === MAX_INT) {
      board[next_x][next_y] = board[cur_x][cur_y] + 1;
      que.push([next_x, next_y]);
    }
  }
}

que.push([x, y, 0]);
visited[x][y] = true;
let is_end = false;

while (que.length > 0) {
  const [cur_x, cur_y, time] = que.shift();

  if (cur_x === 1 || cur_y === 1 || cur_x === R || cur_y === C) {
    console.log(time + 1);
    is_end = true;
    break;
  }

  for (let i = 0; i < 4; i++) {
    const next_x = cur_x + dx[i];
    const next_y = cur_y + dy[i];

    if (!visited[next_x][next_y] && board[next_x][next_y] > time + 1) {
      visited[next_x][next_y] = true;
      que.push([next_x, next_y, time + 1]);
    }
  }
}

if (!is_end) {
  console.log("IMPOSSIBLE");
}
